internal/extension/extension.go

```diff
@@ -57,14 +80,71 @@ func (em *ExtensionManager) checkAgentRunning() {
 		logger.Debug("Will use the API")
 		em.isExtensionRunning = false
 	} else {
-		req, _ := http.NewRequest(http.MethodGet, em.helloRoute, nil)
-		if response, err := em.httpClient.Do(req); err == nil && response.StatusCode == 200 {
-			logger.Debug("Will use the Serverless Agent")
-			em.isExtensionRunning = true
-		} else {
-			logger.Debug("Will use the API since the Serverless Agent was detected but the hello route was unreachable")
-			em.isExtensionRunning = false
+		logger.Debug("Will use the Serverless Agent")
+		em.isExtensionRunning = true
+	}
+}
+
+func (em *ExtensionManager) SendStartInvocationRequest(ctx context.Context, eventPayload json.RawMessage) context.Context {
+	body := bytes.NewBuffer(eventPayload)
+	req, _ := http.NewRequest(http.MethodPost, em.startInvocationUrl, body)
+
+	if response, err := em.httpClient.Do(req); err == nil && response.StatusCode == 200 {
+		// Propagate dd-trace context from the extension response if found in the response headers
+		traceId := response.Header.Values(string(DdTraceId))
+		if len(traceId) > 0 {
+			ctx = context.WithValue(ctx, DdTraceId, traceId[0])
+		}
+		parentId := response.Header.Values(string(DdParentId))
+		if len(parentId) > 0 {
+			ctx = context.WithValue(ctx, DdParentId, parentId[0])
 		}
+		samplingPriority := response.Header.Values(string(DdSamplingPriority))
+		if len(samplingPriority) > 0 {
+			ctx = context.WithValue(ctx, DdSamplingPriority, samplingPriority[0])
+		}
+	}
+	return ctx
+}
+
+func (em *ExtensionManager) SendEndInvocationRequest(ctx context.Context, functionExecutionSpan ddtrace.Span, err error) {
+	// Handle Lambda response
+	lambdaResponse := ctx.Value(DdLambdaResponse)
+	content, responseErr := json.Marshal(lambdaResponse)
+	if responseErr != nil {
+		content, _ = json.Marshal("{}")
+	}
+	body := bytes.NewBuffer(content)
+
+	// Build the request
+	req, _ := http.NewRequest(http.MethodPost, em.endInvocationUrl, body)
+
+	// Mark the invocation as an error if any
+	if err != nil {
+		req.Header[string(DdInvocationError)] = append(req.Header[string(DdInvocationError)], "true")
```

Comment:
Since we're only expecting to include value for each of these header keys, consider using `req.Header.Set` instead.